### ISSUE TITLE
The Euclidean Space is Evil: Hyperbolic Attribute Editing for Few-shot Image Generation

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#194]
# TL; DR

- Encode real images in the hyperbolic space (Poincar disk) to leverage its hierarchical nature for few-shot image generation.

# Motivation

<img width="574" alt="image" src="https://user-images.githubusercontent.com/46308407/204118321-19d0345a-3010-48aa-b3e3-aae307f0ae57.png">

- Hierarchy is a key inductive bias for data-efficient training, and hyperbolic space is best suited for encoding such quality.

# Formulation

### Framework

<img width="1191" alt="image" src="https://user-images.githubusercontent.com/46308407/204118328-323b0c9e-ed89-48bb-868e-7398a8f17a0a.png">

- Take the off-the-shelf GAN and inversion method (e.g., StyleGAN2 and pSp).
- Simply add Mbius linear layer for hyperbolic embedding, along with exponential and logarithmic mapping.

<img width="571" alt="image" src="https://user-images.githubusercontent.com/46308407/204118359-06174c81-f8c4-4487-9d79-e68b23ae4896.png">

- The major benefit of using a hyperbolic space is that as it resembles a continuous tree structure, the average between two leaf nodes yields their parents, rather than another leaf node along the edge.

### Objective

- Supervised loss to embed latent codes to the hyperbolic space using Poincar extension of multinomial logistic regression.

<img width="277" alt="image" src="https://user-images.githubusercontent.com/46308407/204118627-c5f417c1-e86d-46d9-8b65-4390ca6f4588.png">

- Reconstruction loss for the image and the inverted latent code.

<img width="285" alt="image" src="https://user-images.githubusercontent.com/46308407/204118653-1c84f497-facc-41bd-a04c-bb4c61cd3dfe.png">
<img width="390" alt="image" src="https://user-images.githubusercontent.com/46308407/204118658-f37178ff-bcd2-48bc-a88c-3c6529b89c63.png">
<img width="261" alt="image" src="https://user-images.githubusercontent.com/46308407/204118662-54ad7bc0-7ab6-4a59-96f8-3e0aea51ab86.png">

- Overall,
<img width="451" alt="image" src="https://user-images.githubusercontent.com/46308407/204118664-93cb28a2-9e12-4e53-917d-2c77040f11cf.png">

### Image Generation

- Generation is performed by interpolating between two images (which corresponds to the geodesic).
<img width="545" alt="image" src="https://user-images.githubusercontent.com/46308407/204118674-5bdfa049-d9c0-4ede-928c-ced96e5252d0.png">

# Evaluation

<img width="1176" alt="image" src="https://user-images.githubusercontent.com/46308407/204118688-055903dc-437c-4c3c-aa79-6a092ff71e86.png">

- Different radii controls the level of variation (coarse-fine).

<img width="580" alt="image" src="https://user-images.githubusercontent.com/46308407/204118696-1564268d-cb1f-4ce6-b67e-bb30207a6602.png">
<img width="576" alt="image" src="https://user-images.githubusercontent.com/46308407/204118699-8a475d6d-e346-4f37-b7bd-9aa2a16adb3b.png">
<img width="663" alt="image" src="https://user-images.githubusercontent.com/46308407/204118704-a604dc51-c132-478e-b4c3-2cb1487924f2.png">
<img width="325" alt="image" src="https://user-images.githubusercontent.com/46308407/204118710-8e1c0593-eb86-47da-bb6c-d7dda5444d77.png">

# Inspiration

- Hyperbolic representation learning is gaining attention from the community due to its data- and compute- efficient potentials.
- It would have been better if the authors released the code base, as there is a clear shortage of well organized code bases for hyperbolic learning.
- The idea of leveraging the visual hierarchy is persuasive. More possible extension to compositionality.
